From 11f659008536d6888e2321bdc3a099bdefeda9fa Mon Sep 17 00:00:00 2001
From: repo sync -j8 --no-tags -qc <quic_nandhaki@quicinc.com>
Date: Thu, 12 Oct 2023 12:18:16 +0530
Subject: [PATCH] net: skbuff: Clear skb flags during skb copy

Clear fast_xmit and other recycler related flags during
skb_copy.

Change-Id: I0cd3a31132ce1d65899b97129a76aab73e846c38
Signed-off-by: Nandha Kishore Easwaran <quic_nandhaki@quicinc.com>
CRs-Fixed: 3557441
---
 net/core/skbuff.c | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/net/core/skbuff.c b/net/core/skbuff.c
index f18f2ac67c3f..6570658e1666 100644
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -1365,6 +1365,15 @@ static void __copy_skb_header(struct sk_buff *new, const struct sk_buff *old)
 	new->queue_mapping = old->queue_mapping;
 
 	memcpy(&new->headers, &old->headers, sizeof(new->headers));
+	/* Clear the skb recycler flags here to make sure any skb whose size
+	 * has been altered is not put back into recycler pool.
+	 */
+	new->fast_xmit = 0;
+	new->is_from_recycler = 0;
+	new->fast_recycled = 0;
+	new->recycled_for_ds = 0;
+	new->fast_qdisc = 0;
+	new->int_pri = 0;
 	CHECK_SKB_FIELD(protocol);
 	CHECK_SKB_FIELD(csum);
 	CHECK_SKB_FIELD(hash);
@@ -1393,7 +1402,6 @@ static void __copy_skb_header(struct sk_buff *new, const struct sk_buff *old)
 #ifdef CONFIG_NET_SCHED
 	CHECK_SKB_FIELD(tc_index);
 #endif
-
 }
 
 /*
@@ -2132,6 +2140,16 @@ int pskb_expand_head(struct sk_buff *skb, int nhead, int ntail,
 	if (!skb->sk || skb->destructor == sock_edemux)
 		skb->truesize += size - osize;
 
+	/* Clear the skb recycler flags here to make sure any skb whose size
+	 * has been expanded is not put back into recycler.
+	 */
+	skb->fast_xmit = 0;
+	skb->is_from_recycler = 0;
+	skb->fast_recycled = 0;
+	skb->recycled_for_ds = 0;
+	skb->fast_qdisc = 0;
+	skb->int_pri = 0;
+
 	return 0;
 
 nofrags:
-- 
2.34.1

